### PR TITLE
Unit test: Verify hc feature-gate disable after enable scenario

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
This unit test verifies that if hc feature-gate is disabled after enabling it, hc-controller gets removed.

### CODE REVIEW

File: pkg/operator/operator_test.go [hunk 1]
```diff
@@ -208,6 +208,9 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
 			},
 		},
 		expectedMachineHealthCheckController: true,
+	}, {
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: How does this verify the feature is disable after it's enabled? The operator is initialized from scratch for every test case.

Comment (enxebre) [reply pkg/operator/operator_test.go]: indeed, this is duplicating the tc https://github.com/openshift/machine-api-operator/pull/299/files#diff-d5d9cf6e8104e5231787b5cc89fb7b52R198

File: pkg/operator/operator.go [hunk 2]
```diff
@@ -196,37 +206,45 @@ func (optr *Operator) sync(key string) error {
...
+		TargetNamespace: optr.namespace,
+		Controllers: Controllers{
+			Provider:           providerControllerImage,
+			NodeLink:           machineAPIOperatorImage,
+			MachineHealthCheck: machineAPIOperatorImage,
+		},
+	}, nil
+}
+
+func (optr *Operator) getImages() (string, string, error) {
```
Comment (ingvagabund) [inline pkg/operator/operator.go]: No need to implement and mock `getImages()`. You can create `Configv1.Infrastructure` object and have the fake client return it.

File: vendor/github.com/openshift/cluster-version-operator/lib/resourcemerge/core.go [hunk 3]
```diff
@@ -56,6 +56,15 @@ func ensurePodSpec(modified *bool, existing *corev1.PodSpec, required corev1.Pod
 		}
 		ensureContainer(modified, existingCurr, required)
 	}
+	var requiredContainers []corev1.Container
```
Comment (ingvagabund) [inline vendor/github.com/openshift/cluster-version-operator/lib/resourcemerge/core.go]: How is this change related to the unit tests?

File: pkg/operator/operator.go [hunk 4]
```diff
@@ -47,7 +49,11 @@ type Operator struct {
 	osClient      osclientset.Interface
 	eventRecorder record.EventRecorder
 
-	syncHandler func(ic string) error
+	syncHandler           func(ic string) error
+	imageGetterFunc       imageGetterFuncType
+	statusProgressingFunc func() error
```
Comment (ingvagabund) [inline pkg/operator/operator.go]: No need to mock `optr.statusProgressing` given you can have fake clients return expected objects. Please, give fake client `osconfigv1.ClusterOperator` object with pre-populated testing values.

File: vendor/github.com/openshift/cluster-version-operator/lib/resourcemerge/core.go (hunk 3 above)
Comment (vikaschoudhary16) [reply vendor/github.com/openshift/cluster-version-operator/lib/resourcemerge/core.go]: https://github.com/openshift/cluster-version-operator/pull/178

File: pkg/operator/operator.go (hunk 4 above)
Comment (ingvagabund) [reply pkg/operator/operator.go]: Implemented here: https://github.com/openshift/machine-api-operator/pull/304

**Comment by enxebre** [comment]
@vikaschoudhary16 needs rebase

**Comment by vikaschoudhary16** [comment]
This PR still needs https://github.com/openshift/cluster-version-operator/pull/216/files

File: pkg/operator/operator_test.go [hunk 5]
```diff
@@ -217,22 +229,27 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
...
 
 	for _, tc := range tests {
-		infra := &v1.Infrastructure{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: "cluster",
-			},
-			Status: v1.InfrastructureStatus{
-				Platform: v1.AWSPlatformType,
-			},
+		err := initFeatureGateInformer([]runtime.Object{tc.featureGate}, optr)
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: This is still wrong. All tests in the list (`tests`) are supposed to assume clean/vanilla environment. Initializing fg informer only once leaves previous fg object in the cache. Which might (or might not) influence result of each test in the last. If you want to assume dependency between fg objects, you need to create separate unit test for it.

File: pkg/operator/operator_test.go [hunk 6]
```diff
@@ -228,11 +240,59 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
...
+		},
+		Status: v1.InfrastructureStatus{
+			Platform: v1.AWSPlatformType,
+		},
+	}
+
+	stopCh := make(<-chan struct{})
+	optr := newFakeOperator(nil, []runtime.Object{infra}, stopCh)
+
+	for _, tc := range tests {
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: Given the unit test will always have two steps (first enable fg, then disable fg), there is no need to iterate through both tests. It's sufficient to initialize `newFakeOperator` as it was before, remove `initFeatureGateInformer` and just call `optr.sync` two times. After the first call validating the fg is enable, after the second call validating it is not.

File: pkg/operator/operator_test.go [hunk 7]
```diff
@@ -228,11 +240,59 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
...
 			},
 		}
-
 		stopCh := make(<-chan struct{})
-		optr := newFakeOperator(nil, []runtime.Object{tc.featureGate, infra}, stopCh)
-		go optr.Run(2, stopCh)
+		optr := newFakeOperator(nil, []runtime.Object{infra}, stopCh)
+		err := initFeatureGateInformer([]runtime.Object{tc.featureGate}, optr)
+		if err != nil {
+			t.Errorf("Failed to set featuregate %v : %v", tc.featureGate, err)
```
Comment (frobware) [inline pkg/operator/operator_test.go]: Isn't this fatal?

File: pkg/operator/operator_test.go (hunk 6 above)
Comment (vikaschoudhary16) [reply pkg/operator/operator_test.go]: >  remove initFeatureGateInformer

This is not possible because modifying the `FeatureGate` object which is passed to the fake clientset, will not work.  Eventually it makes a copy of the object and therefor any changes to the object will not be reflected back in the object returned by informer in response to the listing operation.

Updated the PR to address other parts. PTAL!

File: pkg/operator/operator_test.go (hunk 7 above)
Comment (vikaschoudhary16) [reply pkg/operator/operator_test.go]: with `Errorf` this test will be reported failed but others will be executed. Is not that we want?

File: pkg/operator/operator_test.go (hunk 6 above)
Comment (ingvagabund) [reply pkg/operator/operator_test.go]:
```

Comment (ingvagabund) [reply pkg/operator/operator_test.go]: @vikaschoudhary16 also please improve the messages to be more human readable

File: pkg/operator/operator_test.go [hunk 8]
```diff
@@ -74,7 +75,8 @@ func newFakeOperator(kubeObjects []runtime.Object, osObjects []runtime.Object, s
 	}
 
 	configSharedInformer.Start(stopCh)
-	kubeNamespacedSharedInformer.Start(stopCh)
+	kubeInformerstopCh := make(<-chan struct{})
```
Comment (frobware) [inline pkg/operator/operator_test.go]: Why do we need this separate channel? Shouldn't the following line use `stopCh`?

File: pkg/operator/operator_test.go [hunk 9]
```diff
@@ -52,7 +53,7 @@ func newOperatorConfig() *OperatorConfig {
 	}
 }
 
-func newFakeOperator(kubeObjects []runtime.Object, osObjects []runtime.Object, stopCh <-chan struct{}) *Operator {
+func newFakeOperator(kubeObjects []runtime.Object, osObjects []runtime.Object, stopCh chan struct{}) *Operator {
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: No need for `s/<-chan/chan/`.

File: pkg/operator/operator_test.go (hunk 6 above)
Comment (vikaschoudhary16) [reply pkg/operator/operator_test.go]: updated, thanks!

File: pkg/operator/operator_test.go (hunk 8 above)
Comment (ingvagabund) [reply pkg/operator/operator_test.go]: Agree, please undo the change.

File: pkg/operator/operator_test.go [hunk 10]
```diff
@@ -157,7 +158,7 @@ func TestOperatorSync_NoOp(t *testing.T) {
 				},
 			}
 
-			stopCh := make(<-chan struct{})
+			stopCh := make(chan struct{})
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: Please undo

File: pkg/operator/operator_test.go [hunk 11]
```diff
@@ -229,7 +230,7 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
 			},
 		}
 
-		stopCh := make(<-chan struct{})
+		stopCh := make(chan struct{})
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: Please undo

File: pkg/operator/operator_test.go [hunk 12]
```diff
@@ -249,3 +250,61 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
...
+	go optr.Run(2, stopCh)
+
+	if err := wait.PollImmediate(1*time.Second, 5*time.Second, func() (bool, error) {
+		d, err := optr.deployLister.Deployments(targetNamespace).Get(deploymentName)
+		if err != nil {
+			t.Logf("Failed to get %q deployment: %v", deploymentName, err)
+			return false, nil
+		}
+		if !deploymentHasContainer(d, hcControllerName) {
+			t.Logf("Expected deploymentHasContainer for %q container to be true", hcControllerName)
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: `Expected to found no %q container in %q deployment, got one.`

File: pkg/operator/operator_test.go [hunk 13]
```diff
@@ -249,3 +250,61 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
...
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		t.Fatalf("Failed to verify %q deployment", deploymentName)
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: "Failed to verify %q deployment has %q container"

File: pkg/operator/operator_test.go [hunk 14]
```diff
@@ -249,3 +250,61 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
...
+	}
+
+	if err := wait.PollImmediate(1*time.Second, 15*time.Second, func() (bool, error) {
+		d, err := optr.deployLister.Deployments(targetNamespace).Get(deploymentName)
+		if err != nil {
+			t.Logf("Failed to get %q deployment: %v", deploymentName, err)
+			return false, nil
+		}
+		if deploymentHasContainer(d, hcControllerName) {
+			t.Logf("Expected deploymentHasContainer for %q container to be false", hcControllerName)
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: `Expected to found %q container in %q deployment, found none.`

File: pkg/operator/operator_test.go [hunk 15]
```diff
@@ -249,3 +250,61 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
...
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		t.Errorf("Failed to verify %q deployment", deploymentName)
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: "Failed to verify %q deployment has no %q container"

File: pkg/operator/operator_test.go [hunk 16]
```diff
@@ -52,14 +52,17 @@ func newOperatorConfig() *OperatorConfig {
 	}
 }
 
-func newFakeOperator(kubeObjects []runtime.Object, osObjects []runtime.Object, stopCh <-chan struct{}) *Operator {
+func newFakeOperator(kubeObjects []runtime.Object, osObjects []runtime.Object, stopCh <-chan struct{}, wq string) *Operator {
```
Comment (frobware) [inline pkg/operator/operator_test.go]: what is `wq`?

Comment (ingvagabund) [reply pkg/operator/operator_test.go]: If this is the way to fix the flake, please open another PR for it. This is irrelevant for this PR.

Comment (vikaschoudhary16) [reply pkg/operator/operator_test.go]: will it be ok if i create a new commit for this flake fix. Actually with the changes in this PR, this flake has become almost 100% reproducible and thus will be difficult to merge this PR without this fix.

File: pkg/operator/operator_test.go (hunk 11 above)
Comment (vikaschoudhary16) [reply pkg/operator/operator_test.go]: why should not the stop channel be closed?

Comment (ingvagabund) [reply pkg/operator/operator_test.go]: undo = `s/<-//`. Channel only for reading is all that is needed.

File: pkg/operator/operator_test.go (hunk 16 above)
Comment (ingvagabund) [reply pkg/operator/operator_test.go]: How can different name for a working queue be a fix for the flake?

Comment (frobware) [reply pkg/operator/operator_test.go]: How does the name, or a different name, fix the flake?

Comment (ingvagabund) [reply pkg/operator/operator_test.go]: Still, it's unrelated to this PR. Please, open another one.
